File: backend/surveys/rich_text.py
import re
from html import escape
from html.parser import HTMLParser

import bleach
from bleach.css_sanitizer import CSSSanitizer

ALLOWED_RICH_TEXT_TAGS = [
    "p",
    "br",
    "strong",
    "em",
    "u",
    "ul",
    "ol",
    "li",
    "span",
]
ALLOWED_RICH_TEXT_COLORS = {
    "#0f172a",
    "#2563eb",
    "#059669",
    "#dc2626",
    "#d97706",
    "#7c3aed",
}
ALLOWED_RICH_TEXT_ALIGNMENTS = {"left", "center", "right"}
ALLOWED_RICH_TEXT_STYLES_BY_TAG = {
    "p": {"text-align"},
    "ul": {"text-align"},
    "ol": {"text-align"},
    "li": {"text-align"},
    "span": {"color"},
}
RICH_TEXT_CSS_SANITIZER = CSSSanitizer(
    allowed_css_properties=["color", "text-align"],
)
RICH_TEXT_CLEANER = bleach.Cleaner(
    tags=ALLOWED_RICH_TEXT_TAGS,
    attributes={
        tag: ["style"] for tag in ALLOWED_RICH_TEXT_STYLES_BY_TAG
    },
    strip=True,
    css_sanitizer=RICH_TEXT_CSS_SANITIZER,
)


def _normalize_color(value):
    candidate = (value or "").strip().lower()
    if candidate in ALLOWED_RICH_TEXT_COLORS:
        return candidate

    rgb_match = re.match(
        r"^rgb\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})\s*\)$",
        candidate,
    )
    if not rgb_match:
        return None

    channels = [max(0, min(255, int(part))) for part in rgb_match.groups()]
    normalized = "#{:02x}{:02x}{:02x}".format(*channels)
    return normalized if normalized in ALLOWED_RICH_TEXT_COLORS else None


def _normalize_style(tag, style_value):
    allowed_properties = ALLOWED_RICH_TEXT_STYLES_BY_TAG.get(tag, set())
    declarations = []

    for declaration in (style_value or "").split(";"):
        if ":" not in declaration:
            continue

        property_name, raw_value = declaration.split(":", 1)
        property_name = property_name.strip().lower()
        raw_value = raw_value.strip()

        if property_name not in allowed_properties or not raw_value:
            continue

        if property_name == "text-align":
            normalized_value = raw_value.lower()
            if normalized_value not in ALLOWED_RICH_TEXT_ALIGNMENTS:
                continue
            declarations.append(f"text-align: {normalized_value}")
            continue

        if property_name == "color":
            normalized_value = _normalize_color(raw_value)
            if not normalized_value:
                continue
            declarations.append(f"color: {normalized_value}")

    return "; ".join(declarations)


def sanitize_rich_text_html(value):
    raw_html = f"{value or ''}".strip()
    if not raw_html:
        return ""

    cleaned_html = RICH_TEXT_CLEANER.clean(raw_html)
    if not cleaned_html:
        return ""

    def _style_replacer(match):
        tag_name = match.group("tag").lower()
        normalized_style = _normalize_style(tag_name, match.group("style"))
        if not normalized_style:
            return f"<{tag_name}>"
        return f'<{tag_name} style="{normalized_style}">'

    cleaned_html = re.sub(
        r'<(?P<tag>p|ul|ol|li|span)\s+style="(?P<style>[^"]*)">',
        _style_replacer,
        cleaned_html,
        flags=re.IGNORECASE,
    )
    cleaned_html = re.sub(
        r"<(?!/?(?:p|br|strong|em|u|ul|ol|li|span)\b)[^>]+>",
        "",
        cleaned_html,
        flags=re.IGNORECASE,
    )

    return cleaned_html if rich_text_to_plain_text(cleaned_html) else ""


class _RichTextPlainTextParser(HTMLParser):
    def __init__(self):
        super().__init__()
        self.parts = []
        self.list_stack = []

    def _append(self, value):
        if value:
            self.parts.append(value)

    def _append_break(self, count=1):
        if not self.parts:
            return

        existing_breaks = 0
        for part in reversed(self.parts):
            if part != "\n":
                break
            existing_breaks += 1

        for _ in range(max(count - existing_breaks, 0)):
            self.parts.append("\n")

    def handle_starttag(self, tag, attrs):
        tag = tag.lower()
        if tag == "br":
            self._append("\n")
            return

        if tag == "ul":
            self.list_stack.append(("ul", 0))
            return

        if tag == "ol":
            self.list_stack.append(("ol", 0))
            return

        if tag == "li":
            if self.parts and self.parts[-1] not in {"\n", " "}:
                self._append("\n")

            if self.list_stack:
                list_type, count = self.list_stack[-1]
                count += 1
                self.list_stack[-1] = (list_type, count)
                prefix = f"{count}. " if list_type == "ol" else "• "
            else:
                prefix = "• "

            self._append(prefix)

    def handle_endtag(self, tag):
        tag = tag.lower()
        if tag == "li":
            self._append("\n")
            return

        if tag in {"p", "ul", "ol"}:
            self._append_break(2 if tag == "p" else 1)

        if tag in {"ul", "ol"} and self.list_stack:
            self.list_stack.pop()

    def handle_data(self, data):
        self._append(data)


def rich_text_to_plain_text(value):
    sanitized_html = f"{value or ''}".strip()
    if not sanitized_html:
        return ""

    parser = _RichTextPlainTextParser()
    parser.feed(sanitized_html)
    parser.close()

    plain_text = "".join(parser.parts)
    plain_text = plain_text.replace("\xa0", " ")
    plain_text = re.sub(r"[ \t]+\n", "\n", plain_text)
    plain_text = re.sub(r"\n[ \t]+", "\n", plain_text)
    plain_text = re.sub(r"\n{3,}", "\n\n", plain_text)
    return plain_text.strip()


def plain_text_to_rich_text_html(value):
    raw_text = f"{value or ''}".replace("\r\n", "\n").strip()
    if not raw_text:
        return ""

    paragraphs = [
        f"<p>{escape(paragraph).replace(chr(10), '<br>')}</p>"
        for paragraph in re.split(r"\n{2,}", raw_text)
        if paragraph.strip()
    ]
    return "".join(paragraphs)

File: frontend/src/utils/richText.js
export const RICH_TEXT_COLOR_PALETTE = [
  { label: 'Slate', value: '#0f172a' },
  { label: 'Blue', value: '#2563eb' },
  { label: 'Green', value: '#059669' },
  { label: 'Red', value: '#dc2626' },
  { label: 'Amber', value: '#d97706' },
  { label: 'Purple', value: '#7c3aed' },
]

const ALLOWED_TAGS = new Set(['p', 'br', 'strong', 'em', 'u', 'ul', 'ol', 'li', 'span'])
const BLOCK_TAGS = new Set(['p', 'ul', 'ol'])
const ALIGNABLE_TAGS = new Set(['p', 'ul', 'ol', 'li'])
const COLORABLE_TAGS = new Set(['span'])
const ALLOWED_COLORS = new Set(RICH_TEXT_COLOR_PALETTE.map((item) => item.value))
const ALLOWED_ALIGNMENTS = new Set(['left', 'center', 'right'])

function canUseDom() {
  return (
    typeof window !== 'undefined' &&
    typeof document !== 'undefined' &&
    typeof DOMParser !== 'undefined'
  )
}

function escapeHtml(value) {
  return `${value ?? ''}`
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&#39;')
}

function normalizeColorValue(value) {
  const candidate = `${value ?? ''}`.trim().toLowerCase()
  if (ALLOWED_COLORS.has(candidate)) {
    return candidate
  }

  const rgbMatch = candidate.match(
    /^rgb\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})\s*\)$/
  )
  if (!rgbMatch) {
    return null
  }

  const channels = rgbMatch.slice(1).map((part) => Math.max(0, Math.min(255, Number(part))))
  const normalized = `#${channels.map((part) => part.toString(16).padStart(2, '0')).join('')}`
  return ALLOWED_COLORS.has(normalized) ? normalized : null
}

function sanitizeElementStyles(element) {
  const tagName = element.tagName.toLowerCase()
  const declarations = []

  if (ALIGNABLE_TAGS.has(tagName)) {
    const textAlign = `${element.style.textAlign || ''}`.trim().toLowerCase()
    if (ALLOWED_ALIGNMENTS.has(textAlign)) {
      declarations.push(`text-align: ${textAlign}`)
    }
  }

  if (COLORABLE_TAGS.has(tagName)) {
    const color = normalizeColorValue(element.style.color)
    if (color) {
      declarations.push(`color: ${color}`)
    }
  }

  if (declarations.length > 0) {
    element.setAttribute('style', declarations.join('; '))
  } else {
    element.removeAttribute('style')
  }
}

function replaceElementTag(element, nextTagName) {
  const replacement = document.createElement(nextTagName)

  Array.from(element.attributes).forEach((attribute) => {
    replacement.setAttribute(attribute.name, attribute.value)
  })

  while (element.firstChild) {
    replacement.appendChild(element.firstChild)
  }

  element.parentNode?.replaceChild(replacement, element)
  return replacement
}

function unwrapElement(element) {
  const parent = element.parentNode
  if (!parent) {
    return
  }

  while (element.firstChild) {
    parent.insertBefore(element.firstChild, element)
  }
  parent.removeChild(element)
}

function wrapRootInlineContent(container) {
  let paragraph = null

  Array.from(container.childNodes).forEach((node) => {
    const isBlockElement =
      node.nodeType === Node.ELEMENT_NODE && BLOCK_TAGS.has(node.tagName.toLowerCase())

    if (isBlockElement) {
      paragraph = null
      return
    }

    const isMeaningfulText = node.nodeType !== Node.TEXT_NODE || `${node.textContent ?? ''}`.trim()
    if (!isMeaningfulText && !paragraph) {
      container.removeChild(node)
      return
    }

    if (!paragraph) {
      paragraph = document.createElement('p')
      container.insertBefore(paragraph, node)
    }

    paragraph.appendChild(node)
  })
}

function normalizeRichTextDom(container) {
  const elements = Array.from(container.querySelectorAll('*'))

  elements.forEach((node) => {
    if (!node.parentNode) {
      return
    }

    const tagName = node.tagName.toLowerCase()
    let element = node

    if (['script', 'style', 'iframe', 'object', 'embed', 'svg', 'meta', 'link'].includes(tagName)) {
      node.remove()
      return
    }

    if (tagName === 'b') {
      element = replaceElementTag(node, 'strong')
    } else if (tagName === 'i') {
      element = replaceElementTag(node, 'em')
    } else if (tagName === 'font') {
      const nextColor = normalizeColorValue(node.getAttribute('color'))
      element = replaceElementTag(node, 'span')
      if (nextColor) {
        element.style.color = nextColor
      }
    } else if (tagName === 'div') {
      element = replaceElementTag(node, 'p')
    } else if (!ALLOWED_TAGS.has(tagName)) {
      unwrapElement(node)
      return
    }

    Array.from(element.attributes).forEach((attribute) => {
      if (attribute.name !== 'style') {
        element.removeAttribute(attribute.name)
      }
    })
    sanitizeElementStyles(element)
  })

  wrapRootInlineContent(container)
}

export function richTextHtmlToPlainText(value) {
  const rawHtml = `${value ?? ''}`.trim()
  if (!rawHtml) {
    return ''
  }

  if (!canUseDom()) {
    return rawHtml.replace(/<[^>]+>/g, '').trim()
  }

  const documentParser = new DOMParser()
  const parsedDocument = documentParser.parseFromString(`<div>${rawHtml}</div>`, 'text/html')
  const container = parsedDocument.body.firstElementChild || parsedDocument.body
  const parts = []

  const append = (nextValue) => {
    if (nextValue) {
      parts.push(nextValue)
    }
  }

  const appendBreak = (count = 1) => {
    if (!parts.length) {
      return
    }

    let trailingBreaks = 0
    for (let index = parts.length - 1; index >= 0; index -= 1) {
      if (parts[index] !== '\n') {
        break
      }
      trailingBreaks += 1
    }

    for (let index = trailingBreaks; index < count; index += 1) {
      parts.push('\n')
    }
  }

  const walk = (node, listContext = null) => {
    if (node.nodeType === Node.TEXT_NODE) {
      append(node.textContent)
      return
    }

    if (node.nodeType !== Node.ELEMENT_NODE) {
      return
    }

    const tagName = node.tagName.toLowerCase()
    if (tagName === 'br') {
      append('\n')
      return
    }

    if (tagName === 'ul' || tagName === 'ol') {
      Array.from(node.childNodes).forEach((childNode, index) =>
        walk(childNode, { listType: tagName, index })
      )
      appendBreak(1)
      return
    }

    if (tagName === 'li') {
      const prefix =
        listContext?.listType === 'ol' ? `${(listContext.index ?? 0) + 1}. ` : '• '
      append(prefix)
      Array.from(node.childNodes).forEach((childNode) => walk(childNode, null))
      append('\n')
      return
    }

    Array.from(node.childNodes).forEach((childNode) => walk(childNode, null))

    if (tagName === 'p') {
      appendBreak(2)
    }
  }

  Array.from(container.childNodes).forEach((node) => walk(node, null))

  return parts
    .join('')
    .replace(/\u00a0/g, ' ')
    .replace(/[ \t]+\n/g, '\n')
    .replace(/\n[ \t]+/g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .trim()
}

export function plainTextToRichTextHtml(value) {
  const normalized = `${value ?? ''}`.replace(/\r\n/g, '\n').trim()
  if (!normalized) {
    return ''
  }

  return normalized
    .split(/\n{2,}/)
    .filter((paragraph) => paragraph.trim())
    .map((paragraph) => `<p>${escapeHtml(paragraph).replaceAll('\n', '<br>')}</p>`)
    .join('')
}

export function sanitizeRichTextHtml(value) {
  const rawHtml = `${value ?? ''}`.trim()
  if (!rawHtml) {
    return ''
  }

  if (!canUseDom()) {
    return plainTextToRichTextHtml(rawHtml.replace(/<[^>]+>/g, ' '))
  }

  const container = document.createElement('div')
  container.innerHTML = rawHtml
  normalizeRichTextDom(container)

  const normalizedHtml = container.innerHTML.trim()
  return richTextHtmlToPlainText(normalizedHtml) ? normalizedHtml : ''
}

export function resolveRichTextEditorHtml(html, plainText = '') {
  const normalizedHtml = sanitizeRichTextHtml(html)
  if (normalizedHtml) {
    return normalizedHtml
  }

  return plainTextToRichTextHtml(plainText)
}

export function getQuestionTextHtml(question) {
  return question?.settings?.rich_text?.text_html ?? ''
}

function mergeQuestionRichTextSettings(settings, textHtml) {
  const nextSettings = {
    ...(settings ?? {}),
  }
  const nextRichText = {
    ...(nextSettings.rich_text ?? {}),
  }

  if (textHtml) {
    nextRichText.text_html = textHtml
  } else {
    delete nextRichText.text_html
  }

  if (Object.keys(nextRichText).length > 0) {
    nextSettings.rich_text = nextRichText
  } else {
    delete nextSettings.rich_text
  }

  return nextSettings
}

export function buildQuestionRichTextUpdate(question, nextHtml) {
  const textHtml = sanitizeRichTextHtml(nextHtml)
  return {
    text: richTextHtmlToPlainText(textHtml),
    settings: mergeQuestionRichTextSettings(question?.settings, textHtml),
  }
}

export function buildMessageDescriptionRichTextValue(messageConfig, nextHtml) {
  const textHtml = sanitizeRichTextHtml(nextHtml)
  const nextConfig = {
    ...(messageConfig ?? {}),
    desc: richTextHtmlToPlainText(textHtml),
  }

  if (textHtml) {
    nextConfig.desc_html = textHtml
  } else {
    delete nextConfig.desc_html
  }

  return nextConfig
}
